### PR TITLE
Fix Lambda VPC config missing for Neptune DB with IAM auth

### PR DESCRIPTION
**Pull request by kirill-stepanishin**
### Description

   Neptune DB clusters always run inside a VPC, but the Lambda resolver was created without VPC configuration when IAM authentication was enabled. The Lambda could not reach the Neptune endpoint, making the `neptune-db + IAM` deployment combination non-functional.

   ### Root Cause

VPC configuration was conditionally applied based on the authentication method (`!NEPTUNE_IAM_AUTH`) rather than the database type (`NEPTUNE_TYPE === NEPTUNE_DB`). This incorrectly treated VPC and IAM as mutually exclusive, when in practice Neptune DB requires VPC connectivity regardless of how authentication is configured.

   ### Changes

   **Pipeline creation** (`src/pipelineResources.js`)
   - `createLambdaRole()`: split if/else into two independent blocks; neptune-db + IAM now gets all three policies (BasicExecution, NeptuneQuery, VPCAccess)
   - `createLambdaFunction()`: gate `VpcConfig` on `NEPTUNE_TYPE === NEPTUNE_DB` instead of `!NEPTUNE_IAM_AUTH`

   **CDK generation** (`src/CDKPipelineApp.js`, `templates/CDKTemplate.js`)
   - Restructure template from two separate `lambda.Function` calls to incremental props building with a single instantiation
   - Gate VPC config on `NEPTUNE_TYPE`, not auth method
   - Move `NEPTUNE_DBSubnetGroup` replacement inside the `neptuneClusterInfo` guard (previously produced string `'null'` for neptune-graph)

   **Cleanup** (`src/pipelineResources.js`)
   - Replace hardcoded policy detach blocks with dynamic loop over `LambdaExecutionPolicy*` keys

   **Error handling** (`src/pipelineResources.js`, `src/CDKPipelineApp.js`)
   - Always `exit(1)` when cluster info fetch fails for neptune-db, regardless of IAM flag (previously silently continued for IAM)
   - Improve error messages: "IAM authentication enabled/disabled" instead of "authentication is set to VPC/IAM"

   **Tests** (5 new files, 24 test cases)
   - Full 4-combination matrix (neptune-db/graph × IAM/no-IAM) for role creation, Lambda VpcConfig, policy cleanup, and CDK template output
   - Edge cases: partial cleanup failures, old resources.json format without Policy3, cluster info fetch failures

### CODE REVIEW

File: src/test/CDKTemplate.test.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+function renderTemplate({ neptuneType, neptuneIAMAuth }) {
+    return templateSource
+        .replace("const NEPTUNE_TYPE = '';", `const NEPTUNE_TYPE = '${neptuneType}';`)
+        .replace("const NEPTUNE_IAM_AUTH = false;", `const NEPTUNE_IAM_AUTH = ${neptuneIAMAuth};`)
+        .replace("const NEPTUNE_DBSubnetGroup = null;", "const NEPTUNE_DBSubnetGroup = 'vpc-123';")
+        .replace("const NEPTUNE_DBSubnetIds = null;", "const NEPTUNE_DBSubnetIds = 'subnet-1,subnet-2';")
+        .replace("const NEPTUNE_VpcSecurityGroupId = null;", "const NEPTUNE_VpcSecurityGroupId = 'sg-123';")
+        .replace("const NEPTUNE_HOST = '';", "const NEPTUNE_HOST = 'db.cluster.us-east-1.neptune.amazonaws.com';")
+        .replace("const NEPTUNE_IAM_POLICY_RESOURCE = '*';", "const NEPTUNE_IAM_POLICY_RESOURCE = 'arn:aws:neptune-db:us-east-1:123:cluster-abc/*';");
```
Comment (kmcginnes) [inline src/test/CDKTemplate.test.js]: Since these types of string replacements tend to be brittle, I suggest throwing an exception, which would fail the test, if the string isn't found. That would force anyone changing that code/template to update the tests.

File: src/test/CDKTemplate.test.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    expect(managedPolicies).toContain('service-role/AWSLambdaVPCAccessExecutionRole');
+});
+
+test('neptune-graph with IAM: IAM policy, no VPC config', () => {
+    const { lambdaProps } = evaluateTemplate({ neptuneType: 'neptune-graph', neptuneIAMAuth: true });
+    expect(lambdaProps.vpc).toBeUndefined();
+    expect(lambdaProps.vpcSubnets).toBeUndefined();
+    expect(lambdaProps.securityGroups).toBeUndefined();
+    expect(lambdaProps.initialPolicy).toBeDefined();
+    expect(lambdaProps.initialPolicy[0].actions).toContain('neptune-graph:connect');
```
Comment (kmcginnes) [inline src/test/CDKTemplate.test.js]: Does this not need the same read query access as `neptune-db`?

```
neptune-graph:ReadDataViaQuery
```

File: src/test/CDKTemplate.test.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+});
+
+test('neptune-graph without IAM: no VPC, no IAM policy', () => {
```
Comment (kmcginnes) [inline src/test/CDKTemplate.test.js]: Is this even possible? I thought Neptune Analytics required IAM always.

File: src/test/createPipeline.test.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
+import { jest } from '@jest/globals';
+import { mockIAMSend, mockLambdaSend, mockAppSyncSend, mockNeptuneSend, mockWriteFileSync, mockReadFileSync, mockExit, ExitCalled, registerPipelineMocks, mockNeptuneClusterResponse } from './helpers/pipelineMocks.js';
+
+registerPipelineMocks();
+
+const originalSetTimeout = globalThis.setTimeout;
```
Comment (kmcginnes) [inline src/test/createPipeline.test.js]: This could probably be a bit cleaner using Jest's built in fake timers.

https://jestjs.io/docs/timer-mocks

File: src/test/CDKPipelineApp.test.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    });
+    mockWriteFile.mockResolvedValue(undefined);
+});
+
+const clusterInfo = {
+    host: 'db.cluster.us-east-1.neptune.amazonaws.com', port: '8182',
+    isIAMauth: true, version: '1.3.0.0', dbSubnetGroup: 'default-vpc-123',
+    dbSubnetIds: ['subnet-1', 'subnet-2'], vpcSecurityGroupId: 'sg-123',
+    iamPolicyResource: 'arn:aws:neptune-db:us-east-1:123:cluster-abc/*',
+};
```
Comment (kmcginnes) [inline src/test/CDKPipelineApp.test.js:55]: This seems to be unused.

Comment (kmcginnes) [reply src/test/CDKPipelineApp.test.js:55]: Perhaps this was meant for a happy path test that seems to be missing? Kiro suggested:

```js
  test('should generate CDK file with VPC config for neptune-db with IAM', async () => {
      mockGetNeptuneClusterDbInfoBy.mockResolvedValue(clusterInfo);

      await createAWSpipelineCDK({ ...baseParams, isNeptuneIAMAuth: true, neptuneType: 'neptune-db' });

      expect(mockWriteFile).toHaveBeenCalled();
      const writtenContent = mockWriteFile.mock.calls[0][1];
      expect(writtenContent).toContain("NEPTUNE_IAM_AUTH = true");
      expect(writtenContent).toContain("NEPTUNE_DBSubnetGroup = 'default-vpc-123'");
      expect(writtenContent).toContain("NEPTUNE_DBSubnetIds = 'subnet-1,subnet-2'");
      expect(writtenContent).toContain("NEPTUNE_VpcSecurityGroupId = 'sg-123'");
  });
```

File: src/test/CDKPipelineApp.test.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    expect(mockProcessExit).toHaveBeenCalledWith(1);
+});
+
+test('should exit on cluster info failure for neptune-db without IAM', async () => {
+    mockGetNeptuneClusterDbInfoBy.mockRejectedValue(new Error('cluster not found'));
+
+    try { await createAWSpipelineCDK({ ...baseParams, isNeptuneIAMAuth: false, neptuneType: 'neptune-db' }); } catch (e) { if (!(e instanceof ExitCalled)) throw e; }
+
+    expect(mockProcessExit).toHaveBeenCalledWith(1);
+});
```
Comment (kmcginnes) [inline src/test/CDKPipelineApp.test.js:80]: Missing two of the test scenarios that you have in `createPipeline.test.js`.

```js
  test('should exit when neptune-db has IAM disabled but IAM flag was provided', async () => {
      mockGetNeptuneClusterDbInfoBy.mockResolvedValue({ ...clusterInfo, isIAMauth: false });

      await expect(createAWSpipelineCDK({ ...baseParams, isNeptuneIAMAuth: true, neptuneType: 'neptune-db' }))
          .rejects.toThrow(ExitCalled);
      expect(mockProcessExit).toHaveBeenCalledWith(1);
  });

  test('should exit when neptune-db has IAM enabled but IAM flag was not provided', async () => {
      mockGetNeptuneClusterDbInfoBy.mockResolvedValue({ ...clusterInfo, isIAMauth: true });

      await expect(createAWSpipelineCDK({ ...baseParams, isNeptuneIAMAuth: false, neptuneType: 'neptune-db' }))
          .rejects.toThrow(ExitCalled);
      expect(mockProcessExit).toHaveBeenCalledWith(1);
  });
```

File: src/test/createPipeline.test.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        expect(policyDoc.Statement[0].Action).toContain('neptune-graph:connect');
+    });
+
+    test('should attach only basic execution role for neptune-graph without IAM', async () => {
+        await createUpdateAWSpipeline({ ...baseParams, isNeptuneIAMAuth: false, neptuneType: 'neptune-graph' });
+
+        const attachCalls = mockIAMSend.mock.calls.filter(([c]) => c._type === 'AttachRolePolicy').map(([c]) => c.input.PolicyArn);
+        expect(attachCalls).toContain('arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole');
+        expect(attachCalls).toHaveLength(2); // basic + invoke policy
+        expect(attachCalls).not.toContain('arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole');
```
Comment (kmcginnes) [inline src/test/createPipeline.test.js]:
```suggestion
        expect(attachCalls).toContain('arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole');
        expect(attachCalls).not.toContain('arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole');
        expect(attachCalls).toHaveLength(2); // basic + invoke policy
```

I would at the very least re-order these so if this test was to fail it would provide a decent message, rather than an unhelpful "expected 2, but got 3". You could also just drop the length check, or check the full list of policies is exactly some specific list.

File: src/test/CDKTemplate.test.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        .replace("const NEPTUNE_IAM_POLICY_RESOURCE = '*';", "const NEPTUNE_IAM_POLICY_RESOURCE = 'arn:aws:neptune-db:us-east-1:123:cluster-abc/*';");
+}
+
+function evaluateTemplate({ neptuneType, neptuneIAMAuth }) {
+    const code = renderTemplate({ neptuneType, neptuneIAMAuth });
+    // Extract from "const lambdaProps = {" up to "echoLambda = new lambda.Function"
+    const match = code.match(/const lambdaProps = \{[\s\S]*?(?=\n\s*echoLambda = new lambda\.Function)/);
+    if (!match) throw new Error('Could not extract lambdaProps block from CDKTemplate.js — check that "const lambdaProps = {" and "echoLambda = new lambda.Function" markers still exist in the template');
+
+    const managedPolicies = [];
```
Comment (kmcginnes) [inline src/test/CDKTemplate.test.js]: Seems unused

File: src/test/removePipeline.test.js [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+import { mockIAMSend, mockLambdaSend, mockAppSyncSend, registerPipelineMocks } from './helpers/pipelineMocks.js';
+
+registerPipelineMocks();
+
+let removeAWSpipelineResources;
+
+describe('removeAWSpipelineResources', () => {
+
+beforeEach(async () => {
+    ({ removeAWSpipelineResources } = await import('../pipelineResources.js?t=' + Date.now()));
```
Comment (kmcginnes) [inline src/test/removePipeline.test.js:12]:
```suggestion
    // Re-import with unique query string to reset module-level mutable state between tests
    ({ removeAWSpipelineResources } = await import('../pipelineResources.js?t=' + Date.now()));
```

File: src/test/createPipeline.test.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+registerPipelineMocks();
+
+const originalSetTimeout = globalThis.setTimeout;
+
+let createUpdateAWSpipeline;
+
+beforeEach(async () => {
+    globalThis.setTimeout = (fn) => originalSetTimeout(fn, 0);
+
+    ({ createUpdateAWSpipeline } = await import('../pipelineResources.js?t=' + Date.now()));
```
Comment (kmcginnes) [inline src/test/createPipeline.test.js:13]:
```suggestion
    // Re-import with unique query string to reset module-level mutable state between tests
    ({ createUpdateAWSpipeline } = await import('../pipelineResources.js?t=' + Date.now()));
```

**Review by kmcginnes — COMMENTED** [review]
Mostly suggested cleanup and minor improvements.

This change does not account for Neptune DB public endpoints, which are now a thing. They require IAM to be enabled, but do not require a VPC. I would assume this is out of scope for this change, but worth noting.

File: src/test/CDKTemplate.test.js (hunk 1 above)
Comment (kmcginnes) [reply src/test/CDKTemplate.test.js]: I recently had to do this.

https://github.com/aws/graph-explorer/pull/1640/changes#diff-0a05b1ab392db457fd15d2ff2f21897444be01c2272d8e28dcdf282d6a4d2575R20

Comment (kirill-stepanishin) [reply src/test/CDKTemplate.test.js]: Done. added a `safeReplace` helper that throws if the placeholder is missing. Thanks for the reference.

File: src/test/CDKTemplate.test.js (hunk 2 above)
Comment (kirill-stepanishin) [reply src/test/CDKTemplate.test.js]: Fixed, both IAM tests now use `toEqual` with all 4 actions.

File: src/test/CDKTemplate.test.js (hunk 3 above)
Comment (kirill-stepanishin) [reply src/test/CDKTemplate.test.js]: True, this test is not very useful. Removed it and equivalents in other test files. Thanks!

File: src/test/CDKPipelineApp.test.js (hunk 5 above)
Comment (kirill-stepanishin) [reply src/test/CDKPipelineApp.test.js:55]: Thank you; added the happy-path test that uses it, including IAM policy resource assertion.

File: src/test/CDKPipelineApp.test.js (hunk 6 above)
Comment (kirill-stepanishin) [reply src/test/CDKPipelineApp.test.js:80]: Added both. Thanks!

File: src/test/CDKTemplate.test.js (hunk 8 above)
Comment (kirill-stepanishin) [reply src/test/CDKTemplate.test.js]: Removed.

File: src/test/createPipeline.test.js (hunk 7 above)
Comment (kirill-stepanishin) [reply src/test/createPipeline.test.js]: Good idea, applied your suggestion. Thank you

File: src/test/removePipeline.test.js (hunk 9 above)
Comment (kirill-stepanishin) [reply src/test/removePipeline.test.js:12]: Applied this wording to re-imports across tests. Thank you

File: src/test/createPipeline.test.js (hunk 4 above)
Comment (kirill-stepanishin) [reply src/test/createPipeline.test.js]: Refactored using `jest.useFakeTimers()`. Thanks!

**Comment by kirill-stepanishin** [comment]
> This change does not account for Neptune DB public endpoints, which are now a thing. They require IAM to be enabled, but do not require a VPC. I would assume this is out of scope for this change, but worth noting.

This change addresses exactly this scenario. Neptune DB public endpoints are still gated by security groups, which typically only allow traffic from within the VPC or specific IPs. A Lambda outside the VPC can't connect unless the security group is opened to all IPs, which we shouldn't encourage. Placing the Lambda in the VPC works for both public and private endpoints.

**Review by kmcginnes — APPROVED** [review]
